### DISCUSSION TITLE
Need help with UEFI Tool

### DISCUSSION THREAD

**Discussion by cloudyysouljah** [terminatorul/NvStrapsReBar#49] [Q&A] [answered]
I want to add nvstrap on my uefi bios, but I have a problem. When i try to add dxe nvstraps i get a error on uefi tool and mmtool, as I understand this happening because I don't have much free space in bios, rebar take 2 kb, when nvstrap about 14 kb. I can fix it? Bios 8 mb
![jHv1bkpg3n](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/1507630d-7136-46ee-9a3e-0548d816b3de)
![k0lpPtuP73](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/af94471f-36e9-4abb-b71a-0afc7e8a1502)
![Ibg5LuBgcI](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/45c52338-9b1b-40d6-ae8c-9fcf12bcc469)

**Comment by Tykwha** [comment]
On the 1st screenshot, Strap driver is not the last one in the volume list, perhaps this causes an error?

**Comment by terminatorul** [comment]
You may get a smaller .ffs file if you manually compile [v0.1](https://github.com/terminatorul/NvStrapsReBar/tree/v0.1) from github.

But make sure this is really the issue, because [v0.1](https://github.com/terminatorul/NvStrapsReBar/tree/v0.1) uses hard-coded values for the GPU BAR0 address on the PCI bus (plus the PCI bridge memory mapped address), that you need to fill in manually into the source code.

Also see the [video guide](https://www.youtube.com/watch?v=ldnC18u8FI0) (thanks to @UnidentifiedTag for providing it).

Maybe you can try an older UEFI image for your motherboard, with a smaller size, just to confirm that size is the problem. Make sure the older version still supports your CPU !

To further decrease the .ffs size, you could manually remove all the IDs from DeviceRegistry.c source file, except the PCI device ID for your GPU

**Comment by cloudyysouljah** [comment]
![QiSGkdBFvY](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/51196451-d1c3-43b9-9cd6-02cd460633b6)
I have this problem all day, but yesterday i got build step and have error on him, I did everything like in the video and guide on github, but got error on bulding .ffs driver with python (bad that I didn't screenshot), I sure that problem in small free space and if I make .ffs smaller I can turn on rebar on my mobo

**Comment by cloudyysouljah** [comment]
If you can help with that (make .ffs smaller 9 kb) I will be very grateful. I can send hex values of my system. I will try tomorrow, but i dont sure about success

**Comment by xCuri0** [comment]
@666cloudyy i saw on win-raid forum that some users delete unwanted modules relating to network boot to free up space for nvme dxe on old motherboards. maybe that can work in this case too

**Reply by cloudyysouljah** [reply]
I wanted to try it, but in the beginning I wanted to make a .ffs module from scratch and only then trying delete the modules
But I'm stuck at the last stage with python script

**Reply by UnidentifiedTag** [reply]
What does the python error say?

**Reply by cloudyysouljah** [reply]
say this errors 
![Avyiv3WpXS](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/760246b0-4a6a-41da-9efc-33296a7caaff)
![EDEZ4gGVqx](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/84fefc21-6ec8-4ceb-88a1-3a671632532b)
![image](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/05ac7319-f572-49f5-988d-9629e850fab5)

**Reply by UnidentifiedTag** [reply]
It’s obvious to me what the problem is, you didn’t follow the instructions exactly on the YouTube video when installing cmake or python, hence the script is giving you errors, you need to add the directory’s of all installed tools, this isn’t something that is magically done for you 

https://youtu.be/ldnC18u8FI0?si=bU75GSRVM0yYu5iJ

Run these commands with the correct location to python, cmake, nasm 

set PATH=%PATH%;%ProgramFiles%\NASM\

set PATH=%PATH%;pathtoyourepythoninstall

set PATH=%PATH%;%ProgramFiles%\pathtoyourecmakeinstall 

Note : it might be a sub directory with in the install, is normally where the .exe is located so python.exe or cmake.exe so look for .exe files with in where the installs were then add that directory to each command example “set PATH=%PATH%;” then you’re path

this should clear some of the errors if not all,

**Reply by UnidentifiedTag** [reply]
To check each programs path has been fixed you run 

nasm --version
Python --version
Cmake --version

@terminatorul could you add the rest, because you’re deleted the compile instructions the compile video doesn’t have the commands now… really on the main GitHub you could of kept both this new method you want everyone to use and the old method and just titled each under what they are

**Reply by UnidentifiedTag** [reply]
The only other advice I can give you @666cloudyy is start again and follow exactly what happens in the video, you’ll have to manually copy by eye each command that is used, because they have disappeared from where they were located

**Comment by cloudyysouljah** [comment]
I was able to advance almost to the end, but still something is going wrong. I have this problem after build .ffs and reflash UEFI, i tested several values in LocalPciGPU.h, but he writes this error everywhere
![VqiIT1XHwA](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/94a20171-6c2d-46f6-a661-55cbdae6f6e9)

**Reply by UnidentifiedTag** [reply]
You’ve not compiled your values correctly, using incorrect number, watch the YouTube video about how to get the correct values make sure to search the full sentences and not just the word,

“vga controller”
“Pci to pci Bridge”
 note the correct one will have you’re manufacturer name next to it

read the featured comment on the video to learn more

**Reply by cloudyysouljah** [reply]
i must use decimal values or hex in localpcigpu?

**Reply by cloudyysouljah** [reply]
it's just that in the video I saw that there were three different options and I didn't understand which one to use

**Reply by UnidentifiedTag** [reply]
You can use any of the 3, try any of the 3, the 1st one is obviously recommended that’s why it’s 1st, it goes in order of best too least best. 

You use the website to calculate the numbers you need to enter https://www.save-editor.com/tools/wse_hex.html


I’d say your problem lays in getting the correct digits using the search option in notepad, After cpuid.

“vga controller”
“Pci to pci Bridge”

Again with the correct values you need to copy you’ll see the manufacturers name listed, amd, intel, Nvidia so on…

one will be the motherboard manufacturer the other you’re gpu manufacturer

**Comment by Black6spdZ** [comment]
what steps can we follow to help isolate the problem if our system fails to post with any size ReBAR set?

**Reply by terminatorul** [reply]
@Black6spdZ Are you compiling v0.1 as well ? Usually your simptoms (fail to POST) would indicate the wrong hard-coded values, like for GPU BAR0 base address, or the bridge address range

**Reply by Black6spdZ** [reply]
I'm the one with the dual xeon Z10PE board and I tried the first iteration of the auto driver unless anything substantial has come since then

**Comment by cloudyysouljah** [comment]
I do not know what I am doing wrong, i tested all 3 variants of localpcigpu in .ffs, with or without uefi patch
![n7KZ2Fh80a](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/e95a28d0-4711-4fe4-9751-313b2e32e620)
Maybe I use wrong values, but I don't know - 
Description			VGA Controller
Location			bus 2 (0x02), device 0 (0x00), function 0 (0x00)
Common header
	Vendor ID		0x10DE (NVIDIA)
	Model ID		0x2188
00 00 00 FA - 4194304000
0xFA000000
PCI registers	
		00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F 
	00	DE 10 88 21 07 00 10 00 A1 00 00 03 10 00 80 00 
	10	00 00 00 FA 0C 00 00 E0 1F 38 00 00 0C 00 00 F0 
	20	1F 38 00 00 01 E0 00 00 00 00 00 00 62 14 97 8D 
	30	00 00 00 00 60 00 00 00 00 00 00 00 28 01 00 00 
	40	62 14 97 8D 00 00 00 00 00 00 00 00 00 00 00 00 
	50	00 00 00 00 01 00 00 00 CE D6 23 00 00 00 00 00 
	60	01 68 C3 C9 08 00 00 00 05 78 80 00 00 00 00 00 
	70	00 00 00 00 00 00 00 00 10 00 12 00 E1 8D 2C 11 
	80	10 29 19 00 03 4D 45 00 40 01 03 11 00 00 00 00 
	90	00 00 00 00 00 00 00 00 00 00 00 00 13 00 04 00 
	A0	00 00 00 00 0E 00 00 00 03 00 1E 00 00 00 00 00 
	B0	00 00 00 00 09 00 14 01 00 00 00 00 00 00 00 00 
	C0	00 00 00 00 00 00 00 00 11 00 05 00 00 00 B9 00 
	D0	00 00 BA 00 00 00 00 00 00 00 00 00 62 14 97 8D 
	E0	62 14 97 8D 03 00 00 00 00 00 00 00 00 00 00 00 
	F0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	100	02 00 81 25 00 00 00 00 00 00 00 00 00 00 00 00 
	110	00 00 00 00 01 00 00 80 00 00 00 00 00 00 00 00 
	120	00 00 00 00 00 00 00 00 04 00 01 42 00 00 00 00 
	130	21 81 07 00 00 00 00 00 00 00 00 00 00 00 00 00 
	140	00 00 00 00 30 00 00 01 00 00 00 00 1B 00 00 BB 
	150	95 01 00 E0 20 71 50 06 00 00 00 00 21 81 07 00 
	160	21 81 05 00 02 80 03 00 02 80 01 00 06 80 1F 00 
	170	06 80 1D 00 00 00 00 00 00 00 00 00 00 00 00 00 
	180	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	190	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1A0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1B0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1C0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1D0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1E0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1F0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00


Description			PCI to PCI Bridge
Location			bus 0 (0x00), device 3 (0x03), function 0 (0x00)
Common header
	Vendor ID		0x8086 (Intel)
	Model ID		0x0E08
	Revision ID		0x04
	PI			0x00
	SubClass		0x04
	BaseClass		0x06
	Cache Line		0x10
	Latency			0x00
	Header			0x81
PCI header
	Primary bus		0x00
	Secondary bus		0x02
	Int. Line		0xFF
	Int. Pin		0x01
PCI capability
	Caps class		Subsystem Vendor
	Caps offset		0x40
	SubVendor ID		0x8086 (Intel)
	SubSystem ID		0x0E08
PCI capability
	Caps class		Message Signalled Interrupts
	Caps offset		0x60
PCI capability
	Caps class		PCI Express
	Caps offset		0x90
	Device type		Root Port of PCI-E Root Complex
	Port			0
	Caps version		2.0
	Physical slot		#0
	Presence detect		yes
	Link width		16x (max 16x)
	Link speed		8.0 GT/s (max 8.0 GT/s)
PCI capability
	Caps class		Power Management
	Caps offset		0xE0
	Caps version		1.2
PCI extended capability
	Caps class		Vendor Specific
	Caps offset		0x100
	Vendor			2
	Size			12
PCI extended capability
	Caps class		0xD
	Caps offset		0x110
PCI extended capability
	Caps class		Advanced Error Reporting
	Caps offset		0x148
PCI extended capability
	Caps class		Vendor Specific
	Caps offset		0x1D0
	Vendor			3
	Size			10
PCI extended capability
	Caps class		Secondary PCI Express
	Caps offset		0x250
PCI extended capability
	Caps class		Vendor Specific
	Caps offset		0x280
	Vendor			5
	Size			24
E0 E0 - 57568
0xE0E0
00 FA 00 FB - 4211145216
0xFB00FA00
PCI registers	
		00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F 
	00	86 80 08 0E 07 04 10 00 04 00 04 06 10 00 81 00 
	10	00 00 00 00 00 00 00 00 00 02 02 00 E0 E0 00 20 
	20	00 FA 00 FB 01 E0 01 F2 1F 38 00 00 1F 38 00 00 
	30	00 00 00 00 40 00 00 00 00 00 00 00 FF 01 00 00 
	40	0D 60 00 00 86 80 08 0E 00 00 00 00 00 00 00 00 
	50	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	60	05 90 02 01 00 00 00 00 00 00 00 00 00 00 00 00 
	70	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	80	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	90	10 E0 42 01 21 80 00 00 00 00 00 00 03 79 7A 00 
	A0	40 00 03 F1 80 25 00 00 C0 07 40 00 00 00 01 00 
	B0	00 00 00 00 BE 13 00 00 00 00 00 00 0E 00 00 00 
	C0	03 00 1E 00 00 00 00 00 00 00 00 00 00 00 00 00 
	D0	00 00 00 00 10 00 00 00 00 00 00 00 00 00 00 00 
	E0	01 00 03 C8 00 00 00 00 00 00 00 00 00 00 00 00 
	F0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	100	0B 00 01 11 02 00 C0 00 07 38 00 00 00 00 00 00 
	110	0D 00 81 14 1F 00 00 00 00 00 00 00 00 00 00 00 
	120	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	130	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	140	00 00 00 00 00 00 00 00 01 00 01 1D 00 00 00 00 
	150	00 00 00 00 30 20 06 00 00 00 00 00 00 20 00 00 
	160	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	170	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	180	91 30 18 00 00 00 00 00 00 10 10 60 00 00 00 00 
	190	04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1A0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1B0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1C0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1D0	0B 00 01 25 03 00 A1 00 01 00 00 00 00 00 00 00 
	1E0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
	1F0	00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

**Reply by UnidentifiedTag** [reply]
You’re copying data that we don’t even use? Which suggests you haven’t watched the video?

**Reply by cloudyysouljah** [reply]
No, I use vendor and model ids, values of bus and device, function, and hex values from pci registers - 4194304000 or
0xFA000000 - GPU BAR0 ADDRESS
4211145216 or 0xFB00FA00 - GPU BASE MEM LIMIT
57568 or 0xE0E0 - GPU IO BASE LIMIT

**Reply by UnidentifiedTag** [reply]
Please copy you’re full cpuid report to a pastebin.com so that the same format the code is uploaded in is kept

**Reply by cloudyysouljah** [reply]
https://pastebin.com/SGWLsCgK

**Reply by UnidentifiedTag** [reply]
```
Description			VGA Controller
Location			bus 2 (0x02), device 0 (0x00), function 0 (0x00)
Common header
	Vendor ID		0x10DE (NVIDIA)
	Model ID		0x2188
	
From PCI registers raw

After hex to dec :

4194304000


Description			PCI to PCI Bridge
Location			bus 0 (0x00), device 1 (0x01), function 0 (0x00)
Common header
	Vendor ID		0x8086 (Intel)
	Model ID		0x0E02
	
From PCI registers

After hex to dec :

240

4213242656 
```

**Reply by UnidentifiedTag** [reply]
The above are the numbers you put in to Visual studios, unless you want me to do that for you as well?

please could @terminatorul  or @xCuri0 check that I gave the right numbers to the end user, otherwise it might be a case of the blind leading the blind 

let us know the results @666cloudyy after I have followed my own YouTube video on an iPhone editing you’re cpuid results

**Reply by cloudyysouljah** [reply]
I tested this values, but I have the same result

**Reply by UnidentifiedTag** [reply]
This will need to be escalated to the next support agent 

thank you for you’re patience

**Comment by terminatorul** [comment]
Hi @666cloudyy 

Your LocalPciGpu.h file should have the values below. I used hexadecimal values, but you can freely use the equivalent decimal values if you want.
```C++
#define TARGET_GPU_PCI_VENDOR_ID        0x10DEu
#define TARGET_GPU_PCI_DEVICE_ID        0x2188u

#define TARGET_GPU_PCI_BUS              0x02u
#define TARGET_GPU_PCI_DEVICE           0x00u
#define TARGET_GPU_PCI_FUNCTION         0x00u

// PCIe config register offset 0x10
#define TARGET_GPU_BAR0_ADDRESS         UINT32_C(0xFA00'0000)               // Should fall within memory range mapped by the bridge

#define TARGET_GPU_BAR1_SIZE_SELECTOR   _4G                                // Desired size for GPU BAR1, should cover the VRAM size

// Secondary bus of the bridge must match the GPU bus
// Check the output form CPU-Z .txt report

#define TARGET_BRIDGE_PCI_VENDOR_ID     0x8086u
#define TARGET_BRIDGE_PCI_DEVICE_ID     0x0E08u

#define TARGET_BRIDGE_PCI_BUS           0x00u
#define TARGET_BRIDGE_PCI_DEVICE        0x03u
#define TARGET_BRIDGE_PCI_FUNCTION      0x00u

// Memory range and I/O port range (base + limit) mapped to bridge
// from CPU-Z .txt report of the bridge and GPU

// PCIe config register offset 0x20
#define TARGET_BRIDGE_MEM_BASE_LIMIT  UINT32_C(0xFB00'FA00)                 // Should cover the GPU BAR0

// PCIe config register offset 0x1C
#define TARGET_BRIDGE_IO_BASE_LIMIT   0xE0E0u
```

About your error about the `wrong PCI location property value`.
Please open Device Manager and read the Location information property of your GPU, looking like this:
![image](https://github.com/terminatorul/NvStrapsReBar/assets/378924/eb947878-b83a-4f03-a32d-0afd010635f1)
but for you, it will be in the language you have configured in Windows

Please manually modify the file DeviceList.cc at the following line:
https://github.com/terminatorul/NvStrapsReBar/blob/v0.1/ReBarState/DeviceList.cc#L236
and change the regular expression there to match your language

![image](https://github.com/terminatorul/NvStrapsReBar/assets/378924/c4e3304a-3c53-4c4f-8916-5915c1fb28ad)

The initial version v0.1 was written for the regular expression to match the English language, but later versions no longer depend on language.

**Reply by cloudyysouljah** [reply]
![devenv_Hlz91RUEJW](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/14c4a2c1-986f-4432-b180-f5dada548c6e)
With English language rebar working, but after i change in regexp on my language i get this

**Reply by UnidentifiedTag** [reply]
Yeah you can’t use any other language than English, with 0.1v, so rebar is working ? 

Not sure what you mean by regexp

**Reply by cloudyysouljah** [reply]
I was in a hurry, it turns out that after choosing the rebar size, I have such an error
![kz1esBFx51](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/47bc9757-cb6a-4a86-a63f-03934c920fe7)
And another question - will I be able to use rebar only if the system is in English?

**Reply by UnidentifiedTag** [reply]
Looks to me like you haven’t disabled csm or enabled above 4g encoding in the bios. go confirm the above options are enabled and disabled respectively, open settings, go to recovery, reboot in to recovery mode, don’t press the reset one, when the computer reboots press more options then uefi / bios menu

I have no idea about the last question maybe you could answer that for us? @terminatorul is the user able to use a different ui language with in windows or is this also messed up in v0.1 ? 

Sleep won’t also work @666cloudyy

**Comment by cloudyysouljah** [comment]
![039LDcEO7y](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/8799089a-82c9-489e-9bf6-68d6898f5e38)
I was finally able to do it.
As I understand it, it doesn't work with another language, but if you set the rebar size at the beginning and then change the language, everything will work

**Reply by UnidentifiedTag** [reply]
So the work around is @666cloudyy

 
1) change language to English
2) open the executable (.exe)
3) make the changes
4) change the language back

correct ?

**Reply by cloudyysouljah** [reply]
yeah

**Reply by terminatorul** [reply]
I am happy things worked for you. You can stil enable ReBAR with other languages, if you use the right regular expression. 

This is why I needed a screenshot with the property value in Device Manager on your system. But anyway a regular expression like the following should work for almost any language:
```
[^0-9]*([0-9]+)[^0-9]+([0-9]+)[^0-9]+([0-9]+)[^0-9]*
```

Remember the hard-coded values that you filled in will change if you change settings in UEFI Setup, or if you change your hardware, like adding a new GPU.

So you should disable ReBAR before making such changes, then find the new hard-coded values after the changes, and enable ReBAR again

**Reply by cloudyysouljah** [reply]
![ft15bBDqwZ](https://github.com/terminatorul/NvStrapsReBar/assets/139892412/5378f21c-6374-4860-9899-f2ee713e8a6e)
I have EFI Error, but all programs show that rebar is on - is this how it should be?

**Reply by terminatorul** [reply, marked as answer]
Yes, the old v0.1 has an issue like this, and other users reported the same output. But it works. The issue was fixed later, unfortunately I do not remember what the problem was in the first place
